src/map/skill.c

```diff
@@ -21927,159 +21927,57 @@ static void skill_validate_skillinfo(struct config_setting_t *conf, struct s_ski
-				if (on)
-					sk->inf2 |= INF2_NO_ENEMY;
-				else
-					sk->inf2 &= ~INF2_NO_ENEMY;
-			} else if (strcmpi(skill_info, "IgnoreLandProtector") == 0) {
-				if (on)
-					sk->inf2 |= INF2_NOLP;
-				else
-					sk->inf2 &= ~INF2_NOLP;
-			} else if (strcmpi(skill_info, "Chorus") == 0) {
-				if (on)
-					sk->inf2 |= INF2_CHORUS_SKILL;
-				else
-					sk->inf2 &= ~INF2_CHORUS_SKILL;
-			} else if (strcmpi(skill_info, "FreeCastNormal") == 0) {
-				if (on)
-					sk->inf2 |= INF2_FREE_CAST_NORMAL;
-				else
-					sk->inf2 &= ~INF2_FREE_CAST_NORMAL;
-			} else if (strcmpi(skill_info, "FreeCastReduced") == 0) {
-				if (on)
-					sk->inf2 |= INF2_FREE_CAST_REDUCED;
-				else
-					sk->inf2 &= ~INF2_FREE_CAST_REDUCED;
-			} else if (strcmpi(skill_info, "ShowSkillScale") == 0) {
-				if (on)
-					sk->inf2 |= INF2_SHOW_SKILL_SCALE;
-				else
-					sk->inf2 &= ~INF2_SHOW_SKILL_SCALE;
-			} else if (strcmpi(skill_info, "AllowReproduce") == 0) {
-				if (on)
-					sk->inf2 |= INF2_ALLOW_REPRODUCE;
-				else
-					sk->inf2 &= ~INF2_ALLOW_REPRODUCE;
-			} else if (strcmpi(skill_info, "HiddenTrap") == 0) {
-				if (on)
-					sk->inf2 |= INF2_HIDDEN_TRAP;
-				else
-					sk->inf2 &= ~INF2_HIDDEN_TRAP;
-			} else if (strcmpi(skill_info, "IsCombo") == 0) {
-				if (on)
-					sk->inf2 |= INF2_IS_COMBO_SKILL;
-				else
-					sk->inf2 &= ~INF2_IS_COMBO_SKILL;
-			} else if (strcmpi(skill_info, "BlockedByStasis") == 0) {
-				if (on)
-					sk->inf2 |= INF2_NO_STASIS;
-				else
-					sk->inf2 &= ~INF2_NO_STASIS;
-			} else if (strcmpi(skill_info, "BlockedByKagehumi") == 0) {
-				if (on)
-					sk->inf2 |= INF2_NO_KAGEHUMI;
-				else
-					sk->inf2 &= ~INF2_NO_KAGEHUMI;
-			} else if (strcmpi(skill_info, "RangeModByVulture") == 0) {
-				if (on)
-					sk->inf2 |= INF2_RANGE_VULTURE;
-				else
-					sk->inf2 &= ~INF2_RANGE_VULTURE;
-			} else if (strcmpi(skill_info, "RangeModBySnakeEye") == 0) {
-				if (on)
-					sk->inf2 |= INF2_RANGE_SNAKEEYE;
-				else
-					sk->inf2 &= ~INF2_RANGE_SNAKEEYE;
-			} else if (strcmpi(skill_info, "RangeModByShadowJump") == 0) {
-				if (on)
-					sk->inf2 |= INF2_RANGE_SHADOWJUMP;
-				else
-					sk->inf2 &= ~INF2_RANGE_SHADOWJUMP;
-			} else if (strcmpi(skill_info, "RangeModByRadius") == 0) {
-				if (on)
-					sk->inf2 |= INF2_RANGE_RADIUS;
-				else
-					sk->inf2 &= ~INF2_RANGE_RADIUS;
-			} else if (strcmpi(skill_info, "RangeModByResearchTrap") == 0) {
-				if (on)
-					sk->inf2 |= INF2_RANGE_RESEARCHTRAP;
-				else
-					sk->inf2 &= ~INF2_RANGE_RESEARCHTRAP;
-			} else if (strcmpi(skill_info, "AllowPlagiarism") == 0) {
-				if (on)
-					sk->inf2 |= INF2_ALLOW_PLAGIARIZE;
-				else
-					sk->inf2 &= ~INF2_ALLOW_PLAGIARIZE;
-			} else if (strcmpi(skill_info, "None") != 0) {
-				ShowWarning("%s: Invalid sub-type %s specified for skill ID %d in %s! Skipping sub-type...\n",
+			int j = 0;
+			ARR_FIND(0, ARRAYLENGTH(info_list), j, strcmpi(skill_info, info_list[j].name) == 0);
```

Comment:
Since we have a validation for invalid types, this array ends up being linear-searched for each key in config, but from my tests it doesn't seem to get too slow (striked times were ignored due to being too far from the average):
|         | Current code  | New code      |
| ------- | ------------- | ------------- |
|         | 0.000418s     | ~~0.000589s~~ |
|         | 0.000436s     | 0.000431s     |
|         | 0.000483s     | 0.000441s     |
|         | 0.000477s     | 0.000445s     |
|         | 0.000443s     | 0.000438s     |
|         | 0.000450s     | 0.000445s     |
|         | ~~0.000567s~~ | 0.000440s     |
|         | ~~0.000731s~~ | 0.000466s     |
|         | 0.000434s     | 0.000470s     |
|         | 0.000415s     | 0.000477s     |
|         |               |               |
| Fastest | 0.000415s     | 0.000431s     |
| Slowest | 0.000477s     | 0.000477s     |

I ran it in WSL2, calculating the time around the function call -- I can try in a real Linux if needed.

Reply:
Linear search on the array shouldn't be very different from the chain of if-else conditions (that has the same length as the array and mostly does the same operations), so indeed I wouldn't expect much of a difference, if at all.

To me it basically looks like the opposite of a loop-unroll optimization, but it doesn't make sense to unroll a loop with this amount of operations, at the risk of typos or swapped values in the long if-else chain, so I like what you did here.

I agree it should be strcmp since the DB keys are meant to be case sensitive

Reply:
Thank you, Haru!

In this case, I will take the chance to change this too